crates/ruff_python_parser/src/semantic_errors.rs

```diff
@@ -33,12 +40,20 @@ pub struct SemanticSyntaxChecker {
     /// Python considers it a syntax error to import from `__future__` after any other
     /// non-`__future__`-importing statements.
     seen_futures_boundary: bool,
+
+    /// The checker is currently in an `async` context: either the body of an `async` function or an
+    /// `async` comprehension.
+    ///
+    /// Note that this should be updated *after* checking the current statement or expression
+    /// because the parent context is what matters.
+    in_async_context: bool,
 }
 
 impl SemanticSyntaxChecker {
-    pub fn new() -> Self {
+    pub fn new(source_type: PySourceType) -> Self {
         Self {
             seen_futures_boundary: false,
+            in_async_context: source_type.is_ipynb(),
```

Comment:
I think it would be useful to add a test case for this. We could add a notebook similar to https://github.com/astral-sh/ruff/blob/27ecf350d821d0fa147bc5fb1388626990477890/crates/ruff_linter/resources/test/fixtures/pylint/await_outside_async.ipynb and a new `#[test]` function that uses two Python version.

Reply:
Either is fine. The inlined version (what you have currently) looks good 👍